📖 CAEP: ControlPlane

detiber opened the pull request:
**What this PR does / why we need it**:

Adds a CAEP for ControlPlane Management

vincepri commented:
/milestone v0.3.0

detiber commented:
/assign @dlipovetsky @randomvariable @chuckha @ncdc @vincepri

dlipovetsky commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,495 @@
+  - Pod based control planes; these can be managed via standard kubernetes objects.
+  - Non-node control planes (i.e. EKS, GKE, AKS); these can be managed via the respective APIs.
+- To manage CA certificates outside of what is provided by Kubeadm bootstrapping
+- To manage etcd clusters in any topology other than “stacked etcd.”
```
nit: replace "smart quotes" with normal quotes. (or just remove quotes entirely, since the term is in the glossary)

dlipovetsky commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,495 @@
+
+- If the referenced InitTemplate does not define external etcd (webhook):
+  - ControlPlane.Spec.Replicas in [1, 3, 5, 7]
+- ControlPlane.Spec.Version != “” (openapi)
```
nit: replace "smart" quotes

dlipovetsky commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,495 @@
+    - Poll on cluster health with the correct replica number
+    - Delete the marked controlplane machine
+- Upgrade of worker nodes is deferred to the MachineDeployment controller
+- Determining if a Machine is “up to date” will be done through the use of an annotation (controlplane.cluster.x-k8s.io/configuration-hash) that is placed on the Machine at creation time. The value of this annotation is generated by computing the Hash of the ControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, ...
```
nit: replace "smart" quotes

dlipovetsky commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,495 @@
+
+## Alternatives
+
+For the purposes of designing upgrades, two existing lifecycle managers were examined in detail: kops and Cloud Foundry Container Runtime. Their approaches are detailed in the accompanying “[Cluster API Upgrade Experience Reports](https://docs.google.com/document/d/1RnUG9mHrS_qmhmm052bO6Wu0dldCamwPmTA1D0jCWW8/edit#)” document.
```
nit: replace "smart" quotes

dlipovetsky commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,495 @@
+
+- Will be used during scaling and upgrade operations.
+- There will also be a signal to indicate a healthcheck that could not be made, e.g. due to a network partition.
+- For stacked control planes, we will present etcd quorum status within the Status.Ready field, and also report the number of active cluster members through “ReadyReplicas”
```
nit: replace "smart" quotes

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,495 @@
+[Kubernetes Control Plane Management Overview](https://docs.google.com/document/d/1nlWKEr9OP3IeZO5W2cMd3A55ZXLOXcnc6dFu0K9_bms/edit#)
+
+## Summary
+
```
Suggest adding an intro, something like:

This proposal outlines adding control plane management to Cluster API....

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,495 @@
+
+## Summary
+
+Management of the control plane is currently handled in Cluster API at the infrastructure provider level (although many providers have converged on using the same implementation), based on the special annotation of standard Cluster API machines which then changes the machine bootstrapping logic within a particular infrastructure provider.
```
I don't follow this sentence. We don't use annotations to mark a machine as a control plane machine (we do have a label, though), and infrastructure providers (generally) don't do anything to change machine bootstrapping logic. (There is some logic, such as adding a machine to a load balancer, that an infrastructure provider may do when it detects a new control plane machine.)

akutz replied:
> although many providers have converged on using the same implementation

I'm not sure to what this is referring. What is the convergence?

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,495 @@
+
+- To establish new resource types for control plane management
+- To support single node and multiple control plane instances
+- To enable declarative orchestrated control plane upgrades
```
Do we need to call out that we want to support replacing control plane machines without performing Kubernetes upgrades (because we need to roll out an OS update)?

jzhoucliqr commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,495 @@
+  name: my-cluster
+  namespace: default
+Spec:
+    clusterNetwork:
```
Is clusterNetwork mainly to generate controlPlane kubeadmConfig? If so, is it a more natural fit to move under ControlPlaneSpec?

dlipovetsky replied:
The cluster network (i.e. the pod network) is used by the data plane, although in some cases, control plane component (kube-controller-manager) needs to know the value, too.

For reference, see https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#pod-network.

jzhoucliqr replied:
yeah, but afaik, the definition here is only used to generate ClusterConfiguration for kubeadmConfig, which only applies to ControlPlane. Only exception is APIServerPort which is used by AWS provider to create elb.

detiber replied:
With the *default* implementations this is only currently used in those places. I think it's a valuable conversation to have if having ClusterNetwork defined as part of the Cluster Spec still has value to the project, I'm not sure I want to take that on as part of this proposal, though.

jzhoucliqr replied:
sure. thanks.

jzhoucliqr commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+    - Delete the marked controlplane machine
+- Upgrade of worker nodes is deferred to the MachineDeployment controller
+- Determining if a Machine is "up to date" will be done through the use of an annotation (controlplane.cluster.x-k8s.io/configuration-hash) that is placed on the Machine at creation time. The value of this annotation is generated by computing the Hash of the ControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, ...
+  - For example, a CVE is patched in containerd and updated images are available. Note that the Kubernetes version is not changed. To trigger an upgrade, the user updates the image in the InfrastructureTemplate (for an AWS cluster, the image is stored in InfrastructureTemplate.Spec.AMI.ID).
```
Just curious how will the hash be calculated. Since infrastructureTemplate is an ObjectReference, so calculating hash  of ControlPlane will need to retrieve the actual object (AWSMachineTemplate) and include the Spec of it?

detiber replied:
@jzhoucliqr Since the Infrastructure templates are expected to be immutable after creation, we only need to rely on the reference. This is similar to how we treat the references in MachineDeployments today.

jzhoucliqr replied:
got it. Thanks.

vincepri commented:
/kind proposal

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+  // +optional
+  UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
+
+  // Total number of fully running machines, including the number of active
```
I'm confused about the etcd portion of this comment - this potentially sounds like we're saying the value here would be 6 for a 3-machine control plane (3 machines + 3 etcd cluster members).

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+  // This is the same as ReadyReplicas, but exposed to be similar to the
+  // status exposed by MachineDeployments and Deployments
+  // +optional
+  AvailableReplicas int32 `json:"readyReplicas,omitempty"`
```
Are we going to include the same logic around waiting an optional period of time before considering a replica available?

detiber replied:
I'm not sure that we need to, unless we have some concern that we could have flapping Nodes, but I don't really anticipate that case to be common enough to warrant it.

ncdc replied:
Ok, so my real question is: do we need both Ready & AvailableReplicas, if they're identical? 😄

detiber replied:
I don't think so, but every time I remove it someone asks for it to be re-added again :Joy:

ncdc replied:
If that was me, please accept my apologies! I'd like to avoid having redundant fields.

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+  Initialized bool `json:"initialized,omitempty”`
+
+  // Ready denotes that the ControlPlane API Server is ready to
+  // recieve requests.
```
```suggestion
  // receive requests.
```

ncdc commented on docs/proposals/20191017-control-plane.md:
````diff
@@ -0,0 +1,494 @@
+      cluster.x-k8s.io/controlplane: "true"
+```
+
+- Generate CA certificates if this is the first control plane machine in the cluster
````
```suggestion
- Generate CA certificates if they do not exist
```
?

ncdc commented on docs/proposals/20191017-control-plane.md:
````diff
@@ -0,0 +1,494 @@
+```
+
+- Generate CA certificates if this is the first control plane machine in the cluster
+- Generate the kubeconfig secret if this is the first control plane machine in the cluster
````
```suggestion
- Generate the kubeconfig secret if it does not exist
```
?

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+  namespace: default
+Spec:
+    clusterNetwork:
+    pods:
```
Indent one level?

ncdc commented on docs/proposals/images/controlplane/controlplane-init-1.plantuml:
```diff
@@ -0,0 +1,53 @@
+
+opt Required only if the object doesn't have a Machine owner reference
+"Cluster API Cluster Controller"-> "Cluster API Cluster Controller":Set Cluster.Spec.ControlPlaneRef -> OwnerReferences[0] to Cluster
+"Cluster API Cluster Controller"->"API Server": Patch KubeadMachineControlPlane
```
```suggestion
"Cluster API Cluster Controller"->"API Server": Patch ControlPlane
```

ncdc commented on docs/proposals/images/controlplane/controlplane-init-2.plantuml:
```diff
@@ -0,0 +1,54 @@
+"ControlPlane Controller"->"API Server": Get Machines maching label selector
+"ControlPlane Controller"<<--"API Server": Response
+
+opt ControlPlane.Spec.Scale >= 0, no existing Machines found
```
```suggestion
opt ControlPlane.Spec.Replicas >= 1, no existing Machines found
```

ncdc commented on docs/proposals/images/controlplane/controlplane-init-3.plantuml:
```diff
@@ -0,0 +1,53 @@
+"ControlPlane Controller"->"API Server": Get Machines maching label selector
+"ControlPlane Controller"<<--"API Server": Response
+
+opt ControlPlane.Spec.Scale >= 0, 1 or more existing Machines found that is "Ready"
```
```suggestion
opt ControlPlane.Spec.Replicas >= 1, 1 or more existing Machines found that is "Ready"
```

ncdc commented on docs/proposals/images/controlplane/controlplane-init-6.plantuml:
```diff
@@ -0,0 +1,54 @@
+"ControlPlane Controller"->"API Server": Get Machines maching label selector
+"ControlPlane Controller"<<--"API Server": Response
+
+opt ControlPlane.Spec.Scale >= 0, all existing Machines "Ready", num Machines < Scale
```
```suggestion
opt ControlPlane.Spec.Replicas >= 1, all existing Machines "Ready", num Machines < Replicas
```

ncdc commented on docs/proposals/images/controlplane/controlplane-init-7.plantuml:
```diff
@@ -0,0 +1,57 @@
+
+note over "ControlPlane Controller": - Process for selecting a Machine to delete is TBD
+
+opt ControlPlane.Spec.Scale >= 0, all existing Machines "Ready", num Machines > Scale
```
```suggestion
opt ControlPlane.Spec.Replicas >= 1, all existing Machines "Ready", num Machines > Replicas
```

ncdc commented on docs/proposals/images/controlplane/controlplane-init-3.plantuml:
```diff
@@ -0,0 +1,53 @@
+"ControlPlane Controller"-> "ControlPlane Controller":Set ControlPlane.Status.Initialized = true
+"ControlPlane Controller"-> "ControlPlane Controller":Set ControlPlane.Status.Phase = "Initialized"
+
+opt num Machines matches Scale, and all Machines "Ready"
```
```suggestion
opt num Machines matches Replicas, and all Machines "Ready"
```

ncdc commented on docs/proposals/images/controlplane/controlplane-init-2.plantuml:
```diff
@@ -0,0 +1,54 @@
+
+opt ControlPlane.Spec.Scale >= 0, no existing Machines found
+
+"ControlPlane Controller"->"API Server": Create KubeadmConfig from ControlPlane.Spec.InitTemplate
```
```suggestion
"ControlPlane Controller"->"API Server": Create KubeadmConfig from ControlPlane.Spec.KubeadmConfigSpec
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+- Allow scale up a control plane with stacked etcd to only 1, 3, 5, or 7, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members).
+- However, allow a control plane using an external etcd cluster to scale up to other numbers such as 2 or 4.
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
+- Scale up operations should not be done in conjunction with an upgrade operation.
```
Do we want to reject updates that mutate replicas & version together? What about replicas + other details (infra template, kubeadm config)?

ncdc commented on docs/proposals/images/controlplane/controlplane-init-6.plantuml:
```diff
@@ -0,0 +1,54 @@
+
+opt ControlPlane.Spec.Scale >= 0, all existing Machines "Ready", num Machines < Scale
+
+"ControlPlane Controller"->"API Server": Create KubeadmConfig from ControlPlane.Spec.JoinTemplate
```
```suggestion
"ControlPlane Controller"->"API Server": Create KubeadmConfig from ControlPlane.Spec.KubeadmConfigSpec
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+    - Update the kubeadm generated config map
+- Scale down a control plane with an external etcd
+  - Same as above minus Etcd management
+- Scale down operations should not be done in conjunction with an upgrade operation, this should not impact manual operations for recovery.
```
Same question as above re rejecting

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+#### Identified features from user stories
+
+1. Based on the function of kubeadm, the control plane provider must be able to scale the number of replicas of a control plane from 1 to X, meeting user stories 1 through 4.
+1. To address user story 5, the control plane provider should provide validation of the number of replicas in a control plane. Where the stacked etcd topology is used (i.e., in the default implementation), the number of replicas should be an odd number, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members). When external etcd is used, any number is va ...
```
```suggestion
1. To address user story 5, the control plane provider must provide validation of the number of replicas in a control plane. Where the stacked etcd topology is used (i.e., in the default implementation), the number of replicas must be an odd number, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members). When external etcd is used, any number is valid.
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+
+1. Based on the function of kubeadm, the control plane provider must be able to scale the number of replicas of a control plane from 1 to X, meeting user stories 1 through 4.
+1. To address user story 5, the control plane provider should provide validation of the number of replicas in a control plane. Where the stacked etcd topology is used (i.e., in the default implementation), the number of replicas should be an odd number, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members). When external etcd is used, any number is va ...
+1. In service of user story 5, the control plane provider should also manage etcd membership via kubeadm as part of scaling up and down.
```
```suggestion
1. In service of user story 5, the control plane provider must also manage etcd membership as part of scaling down (`kubeadm` takes care of adding the new etcd member when joining).
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+1. To address user story 5, the control plane provider should provide validation of the number of replicas in a control plane. Where the stacked etcd topology is used (i.e., in the default implementation), the number of replicas should be an odd number, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members). When external etcd is used, any number is va ...
+1. In service of user story 5, the control plane provider should also manage etcd membership via kubeadm as part of scaling up and down.
+1. The control plane provider should provide indicators of health to meet user story 6 and 10. This should include at least the state of etcd and information about which replicas are currently healthy or not. For the default implementation, health attributes based on artifacts kubeadm installs on the cluster may also be of interest to cluster operators.
+1. The control plane provider should be able to upgrade a control plane’s version of Kubernetes as well as updating the underlying machine image on cloud installations.
```
```suggestion
1. The control plane provider must be able to upgrade a control plane’s version of Kubernetes as well as updating the underlying machine image where applicable (e.g. virtual machine based infrastructure).
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+
+##### Create
+
+- A created ControlPlane should bootstrap a control plane with a given number of replicas
```
```suggestion
- A created ControlPlane must bootstrap a control plane with a given number of replicas
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
+- Creating a ControlPlane with > 1 replicas is equivalent to creating a ControlPlane with 1 replica followed by scaling the ControlPlane to 3 replicas
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
```
```suggestion
- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+
+- Allow scale up a control plane with stacked etcd to only 1, 3, 5, or 7, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members).
+- However, allow a control plane using an external etcd cluster to scale up to other numbers such as 2 or 4.
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
```
```suggestion
- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
```

This stanza exists elsewhere in the doc too - do we need it in both places?

detiber replied:
Not likely.

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+- Scale down operations *could* be blocked based on the Health status
+  - for stacked etcd quorum must be considered
+  - for external etcd only availability of etcd should be considered.
+- Scale to 0 should be rejected for the initial support
```
```suggestion
- Scale to 0 must be rejected
```

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+## Glossary
+
+The lexicon used in this document is described in more detail [here](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/reference/glossary.md).  Any discrepancies should be rectified in the main Cluster API glossary.
+Stacked Control Plane / Stacked etcd: Control plane node where etcd is colocated with the Kubernetes API server, and is running as a static pod.
```
Now that https://github.com/kubernetes-sigs/cluster-api/pull/1624 has been merged, we can remove this in favor of the entry in the glossary doc.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+type ControlPlaneSpec struct {
+    // Number of desired machines. Defaults to 1.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    Replicas int32 `json:"replicas"`
```
Mainly to avoid any potential implicit differences between architectures

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+    Replicas int32 `json:"replicas"`
+
+    // Version defines the desired Kubernetes version.
+    Version string `json:"version"`
```
We're missing the same field in Status to detect changes, or is this handled somewhere else?

detiber replied:
This is handled through the use of the configuration hash.

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+##### Cluster upgrade (using create-swap-and-delete)
+
+- Triggered on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used.
+  - Must be able to trigger an upgrade for the underlying machine image (e.g. nothing in the above list has changed but an upgrade is required)
```
How are we planning to handling the image-only update if, in most cases, the infrastructure provider gathers the image id at runtime?

ncdc replied:
I believe that's what the sub-list that immediately follows this line is addressing?

detiber replied:
We are currently deferring the implementation details, but there are some potential solutions listed and we will update the proposal as we flesh it out a bit more in the implementation phase.

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+
+- Triggered on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used.
+  - Must be able to trigger an upgrade for the underlying machine image (e.g. nothing in the above list has changed but an upgrade is required)
+    - Possible implementations include (not necessary to hammer out exact implementation right now):
```
Given that the above is a `Must` and not future work, this proposal should either contain a spec and implementation details, or remove it from here and move it to future work instead / open an issue

detiber replied:
I'd like to keep the MUST in place as a forcing function to require us to deliver the functionality for v1alpha3, but let's defer actual implementation until later so that we can avoid rushing the design right now and gather further feedback for the UX around this.

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+      - Adding a timestamp field, something like `nextUpgrade` which can be set to a specific time in the future (far future if no upgrades are desired) and an upgrade will run when that timestamp is passed. Good for scheduling upgrades/SLOs & quickly modifying to NOW or yesterday if an upgrade is required immediately. Would potentially update the field value to some interval (2 months from now?)
+- An upgrade will look like this:
+  - Serially go through each control plane replica not at the current config hash
+    - Check health of any replicas for the current config hash
```
Same as above, there is no field or indication how to get or expose health information for a control plane

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+- An upgrade will look like this:
+  - Serially go through each control plane replica not at the current config hash
+    - Check health of any replicas for the current config hash
+    - Uniquely mark the existing replica machine through setting an annotation and persisting the change (controlplane.cluster.x-k8s.io/selected-for-upgrade)
```
The ControlPlane object is under `cluster.x-k8s.io`, should this annotation reflect that as well?

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+    - Poll on health of replicas with the current config hash
+    - Poll on cluster health with the correct replica number
+    - Delete the marked controlplane machine
+- Upgrade of worker nodes is deferred to the MachineDeployment controller
```
This is a UX issue that should be addressed in the Risks and mitigation section

detiber replied:
This is out of scope for the Control Plane management and does not represent any changes to the existing behavior.

vincepri replied:
Let's outline what you just said either in R&M or in the non-goal / future work section

detiber replied:
I'd rather just remove this statement all together if it is causing confusion. There is no risk and mitigation required for something that is out of scope (worker nodes are not part of the control plane). This was previously scoped to the Upgrade CAEP that is no longer around.

ncdc replied:
Agree that it doesn't make sense to put in risks & mitigations. +1 to removing entirely, or, at most, listing it as a non-goal.

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,493 @@
+    - Delete the marked controlplane machine
+- Upgrade of worker nodes is deferred to the MachineDeployment controller
+- Determining if a Machine is "up to date" will be done through the use of an annotation (controlplane.cluster.x-k8s.io/configuration-hash) that is placed on the Machine at creation time. The value of this annotation is generated by computing the Hash of the ControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, ...
+  - For example, a CVE is patched in containerd and updated images are available. Note that the Kubernetes version is not changed. To trigger an upgrade, the user updates the image in the InfrastructureTemplate (for an AWS cluster, the image is stored in InfrastructureTemplate.Spec.AMI.ID).
```
This example is very specific to the AWS provider, and it expects users to specify a specific AMI ID, which is rarely the path an operator would go in. We should address this drawback in the risks and mitigation section

ncdc replied:
I wouldn't consider this a rare example. It might not be what happens the majority of the time, but I do anticipate systems that are managing CAPI to do this.

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,492 @@
+- Upgrade of worker nodes is deferred to the MachineDeployment controller
+- Determining if a Machine is "up to date" will be done through the use of an annotation (controlplane.cluster.x-k8s.io/configuration-hash) that is placed on the Machine at creation time. The value of this annotation is generated by computing the Hash of the ControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, ...
+  - For example, a CVE is patched in containerd and updated images are available. Note that the Kubernetes version is not changed. To trigger an upgrade, the user updates the image in the InfrastructureTemplate (for an AWS cluster, the image is stored in InfrastructureTemplate.Spec.AMI.ID).
+- Where necessary, we will import behaviours from the MachineSet and MachineDeployment controllers.
```
Not sure if this needs to be called out, given that it's an implementation detail that we can't guarantee until development time

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,492 @@
+
+- Will be used during scaling and upgrade operations.
+- There will also be a signal to indicate a healthcheck that could not be made, e.g. due to a network partition.
+- For stacked control planes, we will present etcd quorum status within the Status.Ready field, and also report the number of active cluster members through "ReadyReplicas"
```
```suggestion
- For stacked control planes, we will present etcd quorum status within the Status.Ready field, and also report the number of active cluster members through `ControlPlane.Status.ReadyReplicas`.
```

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,492 @@
+
+##### Adoption of pre-v1alpha3 Control Plane Machines
+
+- Existing Control Plane Machines will need to be updated with labels matching the expected label selector.
```
```suggestion
- Existing control plane Machines will need to be updated with labels matching the expected label selector.
```

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,492 @@
+##### Adoption of pre-v1alpha3 Control Plane Machines
+
+- Existing Control Plane Machines will need to be updated with labels matching the expected label selector.
+- The KubeadmConfigSpec can be re-created from the referenced KubeadmConfigs for the Machines matching the label selector.
```
This is a little confusing, mind to reword a little?

chuckha replied:
hmm, i think i have too much context... can you be specific about which part is confusing so i can wordsmith?

vincepri replied:
Well, reading it I don't immediately understand the `re-created from the referenced KubeadmConfigs`. Recreated for what purpose? Which `KubeadmConfigSpec` are we referring to and which referenced `KubeadmConfigs` ?

chuckha replied:
awesome, thanks for the help, let me see what i can do!

ncdc replied:
We'll take the v1a2 control plane machines, look up their bootstrap refs, and then try to create controlPlane.spec.kubeadmConfigSpec from them.

chuckha replied:
How does this sound?

> - The adoptee's KubeadmConfigSpec will need to be regenerated based on the adopter's KubeadmConfigs.

chuckha replied:
@vincepri bump^

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,492 @@
+- Existing Control Plane Machines will need to be updated with labels matching the expected label selector.
+- The KubeadmConfigSpec can be re-created from the referenced KubeadmConfigs for the Machines matching the label selector.
+  - If there is not an existing initConfiguration/clusterConfiguration only the joinConfiguration will be populated.
+- In v1alpha2, the Cluster API Bootstrap Provider is responsible for generating certificates based upon the first machine to join a cluster. The OwnerRef for these certificates are set to that of the initial machine, which causes an issue if that machine is later deleted. For v1alpha3, control plane certificate generation will be replicated in the ControlPlane provider. Given that for v1alpha2 th ...
```
> control plane certificate generation will be replicated in the ControlPlane provider
This should be outlined in the the behavioral changes section above, it wasn't clear while reading top to bottom

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,492 @@
+- The KubeadmConfigSpec can be re-created from the referenced KubeadmConfigs for the Machines matching the label selector.
+  - If there is not an existing initConfiguration/clusterConfiguration only the joinConfiguration will be populated.
+- In v1alpha2, the Cluster API Bootstrap Provider is responsible for generating certificates based upon the first machine to join a cluster. The OwnerRef for these certificates are set to that of the initial machine, which causes an issue if that machine is later deleted. For v1alpha3, control plane certificate generation will be replicated in the ControlPlane provider. Given that for v1alpha2 th ...
+- To ease the adoption of v1alpha3, the migration mechanism should be built into Cluster API controllers.
```
This is very light and needs more information how we're going to migrate, or leave it as an issue

vincepri replied:
One example is spreading the control plane machines across AZs, which is not going to be possible using the implementation outlined in this proposal

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,492 @@
+
+#### Code organization
+
+This proposal puts the types required in the `cluster.x-k8s.io` group. The controller(s) will also live inside `sigs.k8s.io/cluster-api`.
```
```suggestion
The types introduced in this proposal will live under the `cluster.x-k8s.io` API group. The controller(s) will also live inside `sigs.k8s.io/cluster-api`.
```

vincepri commented on docs/proposals/images/controlplane/controlplane-init-2.plantuml:
```diff
@@ -0,0 +1,54 @@
+"ControlPlane Controller"<<--"API Server": Response
+
+"ControlPlane Controller"->"API Server": Create Machine using refs from created KubeadmConfig and InfrastructureMachine resources
+"ControlPlane Controller"<<--"API Server": Response
```
I assume this is done for each Machine, if so can we put this in a loop box?

detiber replied:
I specifically did the diagram around a single init to avoid confusion with introducing loops and branches. This is also why I called out that creating a control plane with replicas > 1 is the same as creating a control plane with a single replica followed by a scale operation.

vincepri replied:
Got it, makes sense it wasn't clear though reading it through

vincepri commented on docs/proposals/images/controlplane/controlplane-init-3.plantuml:
```diff
@@ -0,0 +1,53 @@
+
+' -- GROUPS END ---
+
+note right of "ControlPlane Controller":Watches ControlPlane,\nCluster, and Machine events with mapFunc
```
See above suggestion re: owned machines

vincepri commented on docs/proposals/images/controlplane/controlplane-init-6.plantuml:
```diff
@@ -0,0 +1,54 @@
+"ControlPlane Controller"->"API Server": Get Cluster
+"ControlPlane Controller"<<--"API Server": Response
+
+note over "ControlPlane Controller": - ✅ Cluster.Status.InfraStructureReady is true\n- ✅ ControlPlane instance is valid
```
```suggestion
note over "ControlPlane Controller": - ✅ Cluster.Status.InfrastructureReady is true\n- ✅ ControlPlane instance is valid
```

vincepri commented on docs/proposals/images/controlplane/controlplane-init-7.plantuml:
```diff
@@ -0,0 +1,57 @@
+"ControlPlane Controller"->"API Server": Get Cluster
+"ControlPlane Controller"<<--"API Server": Response
+
+note over "ControlPlane Controller": - ✅ Cluster.Status.InfraStructureReady is true\n- ✅ ControlPlane instance is valid
```
```suggestion
note over "ControlPlane Controller": - ✅ Cluster.Status.InfrastructureReady is true\n- ✅ ControlPlane instance is valid
```

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+type ControlPlaneSpec struct {
+    // Number of desired machines. Defaults to 1.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    Replicas int32 `json:"replicas"`
```
The pointer issue has already been resolved.

The purpose of the definitions here are more around discussion rather than canonical implementation. We can always discuss additional validation as part of the implementation.

vincepri replied:
From a new user perspective, and as a new reader that assumption wasn't clear and I reviewed these structs were close to being final.

If these are just for discussion purposed, either remove them altogether to avoid confusion and only describe the most important fields, or let's make these complete.

detiber replied:
Validation for this field will need to be done using a validation webhook, since enforcement depends on the configuration provided.

detiber replied:
Added the validation requirements below. We can determine openapi vs webhook validations as part of the implementation.

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+  // describe.. The string will be in the same format as the query-param syntax.
+  // More info about label selectors: http://kubernetes.io/docs/user-guide/labels#label-selectors
+  // +optional
+  Selector string `json:"selector,omitempty"`
```
Required for CRD scaling

detiber replied:
It's a requirement for the kubebuilder scale subresource: https://book.kubebuilder.io/reference/markers/crd.html

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+
+With the following validations:
+
+- If the referenced InitTemplate does not define external etcd (webhook):
```
It's part of the kubeadm config

detiber replied:
We have a copy of the KubeadmConfig used for bootstrapping, updating this to remove the old reference to InitTemplate in favor of `ControlPlane.Spec.KubeadmConfigSpec`, though.

vincepri replied:
Sure, but that should be more clear in this proposal somehow. The kubeadm config can also change as outlined below, which would induce weird behaviors.

detiber replied:
Added a statement that the KubeadmConfigSpec needs to be treated immutably and added some clarification that we are determining etcd usage by introspecting the provided KubeadmConfigSpec.

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+  selector:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: my-cluster
+      cluster.x-k8s.io/controlplane: "true"
```
Can you describe a time when there'd be a machine with this label but it's not owned by CAPI?

vincepri replied:
What if users might want to manage control plane machines via external bootstrap methods? Or as a single machines like in v1alpha2?

detiber replied:
Why? If someone is running a manual control plane today they wouldn't have a ControlPlane resource defined. It seems like adding an additional `owned` label would just complicate the adoption scenario.

ncdc replied:
If they are creating actual `Machine` instances to be used as control plane members, starting in v1a3 they must use the `ControlPlane` to do so. We will no longer support other methods. If they want to manage the control plane outside of CAPI (e.g. `kops` phased adoption), that is totally fine and isn't impacted by this.

vincepri replied:
This assumption and breaking change (which I'm ok with) should be clear in the proposal, is it outlined anywhere?

In addition, the current proposal doesn't support AZ-aware control planes, as outlined below. If we don't solve that use case, I think we shouldn't disallow to use plain Machine objects as control planes.

ncdc replied:
Maybe my `must` is too strong? That's my preference, at least. But I recognize it might be biting off too much for alpha3.

vincepri replied:
Yeah, I'm trying to understand the expectations. 

One clarification about my older comment, by "external bootstrap" I meant other bootstrap providers not living in CAPI that generate `Bootstrap.Data`. The `cluster.x-k8s.io/controlplane: "true"` label we're re-using here, is used to allow proper deletion for example, which we might need to retain.

detiber replied:
If they are using a different bootstrapper, I wouldn't expect them to use the default implementation of ControlPlane (or they would use an implementation of ControlPlane that works with their bootstrapper).

For AZ aware control planes I would expect us to lean on the proposal around representing failure domains rather than try to define it in this proposal: https://github.com/kubernetes-sigs/cluster-api/issues/46

I'll work on getting that incorporated into the doc later.

vincepri replied:
> If they are using a different bootstrapper, I wouldn't expect them to use the default implementation of ControlPlane (or they would use an implementation of ControlPlane that works with their bootstrapper).

That's fine, but the use of the label still applies, which we expect to allow for proper deletion and control plane init.

> For AZ aware control planes I would expect us to lean on the proposal around representing failure domains rather than try to define it in this proposal: #46

Sounds good to me, thanks!

akutz replied:
> If they are using a different bootstrapper, I wouldn't expect them to use the default implementation of ControlPlane 

Based on the discussion above I'm concerned we're starting to create a series of disparate components (control plane controller, bootstrapped controller) that all work together well when using the reference implementations of each, but start to fall down when we start picking and replacing one or the other. 

Hmmm. I wonder if we should consider bundling these providers under a single aegis. Not just repo, but making them part of the CAPI controller manager and making it a conscious choice to disable them when configuring CAPI.

detiber replied:
CABPK is already part of the CAPI controller manager for v1alpha3, this controller will also be as well.

That said, you are raising a concern that I have around the interoperability of components when "swapping the batteries" since we will now be having things under the same tree.

akutz replied:
> CABPK is already part of the CAPI controller manager for v1alpha3

Got it.

> That said, you are raising a concern that I have around the interoperability of components when "swapping the batteries" since we will now be having things under the same tree.

Exactly. I don't really see any value in these being modular components if they're mutually modular. At some point the value of modularity seems to only what we're saying, not how we're actually implementing things.

ncdc replied:
I believe there is still value in this modularity. And it's ok that we have default implementations for control plane and bootstrapper that go together. The key value proposition is the generic linkage from `Cluster` to some control plane resource via `.spec.controlPlaneRef`, and from `Machine` to some bootstrap config resource via `.spec.bootstrap.configRef`. These are not prescriptive, and the contracts we've defined and are defining for control plane and bootstrap resources still hold (e.g. they must have `.status.ready`, `.status.errorReason`, and the like).

I am still optimistic that we'll see other bootstrap and control plane providers and that they'll fit in nicely with the aforementioned contracts.

pablochacin replied:
> If they are using a different bootstrapper, I wouldn't expect them to use the default implementation of ControlPlane (or they would use an implementation of ControlPlane that works with their bootstrapper).

@detiber Sorry if I'm asking an obvietey but, what exactly would prevent this to work? What specific assumptions are made about the bootstrapper? I understand it it kubeadm bootstrapping configuration.

detiber replied:
> @detiber Sorry if I'm asking an obvietey but, what exactly would prevent this to work? What specific assumptions are made about the bootstrapper? I understand it it kubeadm bootstrapping configuration.

In addition to the embedded types that are specific to the Kubeadm bootstrapper, there are going to be assumptions made in the ways that the controller interacts with the components based on the way that kubeadm configures clusters. 

For example, the way that we'll manage etcd will be executing commands through the api server against pods named using the conventions of kubeadm, and referencing file paths that are used by kubeadm. We'll also be modifying a ConfigMap that is added and managed by kubeadm that is referenced when joining additional control plane nodes. As we continue to define the health checks, they will be highly dependent on the way that kubeadm configures it's static pods.

detiber replied:
I'm going to defer adding an `owned` label for now and we can revisit as part of the implementation if needed.

ncdc commented on docs/proposals/images/controlplane/controlplane-init-1.plantuml:
```diff
@@ -0,0 +1,53 @@
+
+opt Required only if the object doesn't have a Cluster owner reference
+"Cluster API Cluster Controller"-> "Cluster API Cluster Controller":Set Cluster.Spec.InfrastructureRef -> OwnerReferences[0] to Cluster
+"Cluster API Cluster Controller"->"API Server": Patch AWSInfrastructureConfig
```
I'm guessing `AWSCluster`?

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+    // This is a pointer to distinguish between explicit zero and not specified.
+    Replicas int32 `json:"replicas"`
+
+    // Version defines the desired Kubernetes version.
```
Sure, but seems worthwhile to add it here so it can closely reflect what the implementation PR will look like, more or less?

detiber replied:
I think this should be out of scope for now, since we don't currently enforce any type of version validation, and anything we introduce could potentially affect the adoption of existing clusters as part of the upgrade.

This is probably something that should be a separate issue, since it affects the current uses of the exisiting version fields we have today.

vincepri replied:
Opened https://github.com/kubernetes-sigs/cluster-api/issues/1629

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
+- Creating a ControlPlane with > 1 replicas is equivalent to creating a ControlPlane with 1 replica followed by scaling the ControlPlane to 3 replicas
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
```
Updating, but leaving the *should* in place, since requiring `InitConfiguration` and `ClusterConfiguration` could break adoption scenarios. It also potentially limits the ability to do better defaulting down the line.

jzhoucliqr commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+    - Provision a new machine at the correct version
+    - Poll on health of replicas with the current config hash
+    - Poll on cluster health with the correct replica number
+    - Delete the marked controlplane machine
```
Is the etcd membership management mentioned in "Scale Down" section also needed here?

detiber replied:
Added

akutz commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+
+## Motivation
+
+### Goals
```
I do not see in goals or non-goals any mention of upgrades of single node control planes where there is no external control plane endpoint (CPE). In a Slack conversation with @ncdc, I said my understanding was that single-node control planes without an external CPE would not be supported by any upgrade logic built into CAPI. I think it's worth calling that out explicitly.

akutz commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+### Goals
+
+- To establish new resource types for control plane management
+- To support single node and multiple control plane instances
```
Is single node support dependent upon the existence of an external control plane endpoint?

detiber replied:
Yes, will clarify

detiber replied:
I've updated the goals for this.

akutz commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+    // Phase represents the current phase of control plane actuation.
+    // E.g. Pending, Initialized, Ready, ScalingUp, ScalingDown, Failed etc.
+    // +optional
+    Phase string `json:"phase,omitempty"`
```
The word _phase_ sounds wonky here. Maybe _state_? It works better with well-defined states as well:

```golang
// ControlPlaneState represents the current phase of control plane actuation.
type ControlPlaneState string

const (
	ControlPlanePending     ControlPlaneState = "pending"
	ControlPlaneInitialized ControlPlaneState = "initialized"
	ControlPlaneReady       ControlPlaneState = "ready"
	ControlPlaneScalingUp   ControlPlaneState = "scaling-up"
	ControlPlaneScalingDown ControlPlaneState = "scaling-down"
	ControlPlaneFailed      ControlPlaneState = "failed"
)
```

However, the Kubernetes API conventions indicates [_phase_ is being deprecated in favor of conditions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) anyway:

> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead. 

How would we rewrite the above as conditions?

vincepri replied:
> The word phase sounds wonky here. Maybe state?

We've discussed this during the v1alpha2 planning, `phase` is the preferred word to describe the state the object is in,  I'd prefer if we keep things consistent between our types.

 > How would we rewrite the above as conditions?

This is a larger conversation on how we define and describe a state machine. @timothysc is thinking about this problem already, I'd suggest to move this topic to an issue though.

detiber replied:
This is something that we discussed earlier on the Google doc and decided to stick with Phase to align with the other Cluster API components for now, and we should attempt to address this holistically in the future. /cc @timothysc

akutz replied:
Isn't it exactly the time to adopt recommendations rather than defer, defer, defer? Just because it's hard doesn't mean it's wrong. From what I read _conditions_ are the recommended approach, and it seems like we should figure out how to use them since this is net-new work.

randomvariable replied:
We also have this note that says Conditions themselves are being deprecated https://github.com/detiber/cluster-api/blob/master/docs/proposals/machine-api-proposal.md#conditions

randomvariable replied:
Believe this is the history of the statements: https://github.com/kubernetes/kubernetes/issues/51594#issuecomment-334646068

It's a case not to base things not on Deployments in the core API, but on the newer StatefulSets and DaemonSets instead.

detiber replied:
Considering the contention this topic brings, I'd like to defer this to the issue created to track this discussion more broadly: https://github.com/kubernetes-sigs/cluster-api/issues/1658

akutz commented on docs/proposals/20191017-control-plane.md:
````diff
@@ -0,0 +1,544 @@
+```go
+    // ControlPlaneReady defines if the control plane is ready
+    // +optional
+    ControlPlaneReady bool `json:"controlPlaneReady"`
````
Kubernetes API conventions [highly recommends](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) using `// +optional` where `omitempty` is present

akutz commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+- A created ControlPlane should bootstrap a control plane with a given number of replicas
+- If an error occurs, errorStatus and errorReason will be defined on the ControlPlane object
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
```
Is that important in terms of status? One of the additional goals is:

> - To support pre-existing, user-managed external etcd clusters

But it doesn't indicate that support for external etcd systems includes reporting on the health or status of these external resources.

Although one could argue this point **does** link the two:

>  This should include at least the state of etcd and information about which replicas are currently healthy or not.

So if "state of etcd" is desired and "external etcd" is supported, then one could argue whether etcd is external or not should be part of status.

detiber replied:
I agree that we should do some level of checking even for external etcd, however I'd like to defer that until we get into the implementation rather than fully hash it out now.

akutz commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+- If an error occurs, errorStatus and errorReason will be defined on the ControlPlane object
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
+- Creating a ControlPlane with > 1 replicas is equivalent to creating a ControlPlane with 1 replica followed by scaling the ControlPlane to 3 replicas
```
I don't know, I think this is an important detail. CAPV relies on this today to know that the initial control plane node comes online and is reconciled with the core controllers before the bootstrap provider will create bootstrap data for additional nodes, control plane or otherwise.

I think bringing up an initial node by itself is an important behavior that should be maintained.

detiber replied:
The main purpose of this, in addition to calling out support for the use case @akutz mentions, is that this ties into the provided diagrams, which are simplified to show creation as just bringing up a single control plane machine, and the additional machines are brought up using the scale up workflow.

vincepri replied:
This is fine to keep, I wanted to call out that this stands out because it dictates an internal detail, while the others are worded as simple requirements

akutz replied:
Perhaps it can be reworded to make it more clear that it is a required behavior?

detiber replied:
I'm going to defer this to a later iteration, currently this is mainly to simplify the proposal and related diagrams rather than a long term requirement on the controller.

akutz commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
+- Creating a ControlPlane with > 1 replicas is equivalent to creating a ControlPlane with 1 replica followed by scaling the ControlPlane to 3 replicas
+- The kubeadm bootstrapping configuration provided via `ControlPlane.Spec.KubeadmConfigSpec` should specify the `InitConfiguration`, `ClusterConfiguration`, and `JoinConfiguration` stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate.
```
Is this any different than how this is handled today? If not, why call it out explicitly?

detiber replied:
Today each Machine has a reference to it's own KubeadmConfig, in this case we are taking a singular KubeadmConfig and using it to create the individual KubeadmConfigs for each Machine similar to how the MachineDeployment uses a KubeadmConfigTemplate.

akutz commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+metadata:
+  name: my-cluster
+  namespace: default
+Spec:
```
`s/Spec/spec/`

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+    InfrastructureTemplate corev1.ObjectReference `json:"infrastructureTemplate"`
+
+    // KubeadmConfigSpec is a KubeadmConfigSpec
+    // to use for initializing and joining machines to the control plane
```
Do you have any suggestions?

vincepri replied:
Below for example we suggest that some fields are required, init/join/cluster configuration for example. It'd be great to give some hints on how to use this field as a user. But this can come later in docs.

ncdc replied:
I think we can defer fleshing this out until we have a prototype

akutz commented:
Hi @detiber,

I noticed there are six PNG files in this PR. Since PNG is binary and does not compress well with Git, may I recommend that SVGs are used instead? It is now possible to display SVGs from repo X in GH markdown in repo X with the query string `?sanitize=true` at the end of the SVG's raw path. 

For example, see [this Gist](https://gist.github.com/akutz/d29031271a6d848e2aaaf7350b5febc8) I created with the following markdown:

```markdown
The following SVG won't display because the image link is missing `?sanitize=true`

![controlplane-init-1-nay](https://gist.githubusercontent.com/akutz/d29031271a6d848e2aaaf7350b5febc8/raw/65ebeeff201d6ca7f195e3f22b07461fba4e37a4/controlplane-init-1.svg)

However, when `?sanitize=true` is added, the SVG displays just fine :)

![controlplane-init-1-yay](https://gist.githubusercontent.com/akutz/d29031271a6d848e2aaaf7350b5febc8/raw/65ebeeff201d6ca7f195e3f22b07461fba4e37a4/controlplane-init-1.svg?sanitize=true)
```

![image](https://user-images.githubusercontent.com/101085/67294057-eda2d500-f4aa-11e9-84f3-349a6711b06f.png)

akutz commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+#### Additional goals of the default kubeadm machine-based Implementation
+
+- To provide a kubeadm-based implementation that is infrastructure provider agnostic
+- To manage kubeadm-based "stacked etcd" control plane deployments
```
There should be a comma after _kubeadm-based_ as both it and "stacked etcd" are adjectives describing the type of control plane deployment

akutz commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+- To manage etcd clusters in any topology other than stacked etcd.
+- To address disaster recovery constraints, e.g. restoring a control plane from 0 replicas using a filesystem or volume snapshot copy of data persisted in etcd.
+- To support rollbacks, as there is no datastore rollback guarantee for Kubernetes. Consumers should perform backups of the cluster prior to performing potentially destructive operations.
+- To mutate the configuration of live running clusters (e.g. changing api-server flags), as this is the responsibility of the [component configuration working group](https://github.com/kubernetes/community/tree/master/wg-component-standard).
```
There should be a comma after _live_ as both it and _running_ are adjectives describing _clusters_

chuckha commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,564 @@
+
+## Summary
+
+This proposal outlines a new process for Cluster API to manage control plane machines as a single concept. This includes
```
cc @ncdc 

i reworked the summary and motivation section. ptal!

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+- Allow scale up a control plane with stacked etcd to only 1, 3, 5, or 7, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members).
+- However, allow a control plane using an external etcd cluster to scale up to other numbers such as 2 or 4.
+- The kubeadm bootstrapping configuration provided via ControlPlane.Spec.KubeadmConfigSpec, should specify the initConfiguration, clusterConfiguration, and joinConfiguration stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate
+- Scale up operations should not be done in conjunction with an upgrade operation.
```
Changed to must and added clarification that we expect to enforce it.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+    - Update the kubeadm generated config map
+- Scale down a control plane with an external etcd
+  - Same as above minus Etcd management
+- Scale down operations should not be done in conjunction with an upgrade operation, this should not impact manual operations for recovery.
```
Changed to must and added clarification that we expect to enforce it.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+
+![controlplane-init-7](images/controlplane/controlplane-init-7.png)
+
+##### Delete of the entire ControlPlane (kubectl delete controlplane my-controlplane)
```
Added a statement that user docs should be prioritized towards `kubectl delete cluster` and that Cluster deletion behavior will need to be modified.

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+- To manage CA certificates outside of what is provided by Kubeadm bootstrapping
+- To manage etcd clusters in any topology other than stacked etcd.
+- To address disaster recovery constraints, e.g. restoring a control plane from 0 replicas using a filesystem or volume snapshot copy of data persisted in etcd.
+- To support rollbacks, as there is no datastore rollback guarantee for Kubernetes. Consumers should perform backups of the cluster prior to performing potentially destructive operations.
```
I don't think we necessarily need to provide specific guidance/links on cluster backups in this proposal

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+
+##### Delete of the entire ControlPlane (kubectl delete controlplane my-controlplane)
+
+- Completely removing the control plane and issuing a delete on the underlying machines.
```
Added a comment that Cluster deletion behavior will need to be modified.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+
+## Proposal
+
+### User Stories
```
Updated with explicit numbering

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,544 @@
+type ControlPlaneSpec struct {
+    // Number of desired machines. Defaults to 1.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    Replicas *int32 `json:"replicas"`
```
> I'm very concerned about a situation of how a control plane with static IPs is supposed to exist with the control plane management taking on a model similar to a MachineDeployment with one or more replicas.

Is this something that you believe needs to be solved by the initial implementation? Or is it something that we can likely add through iteration later?

akutz replied:
Hi @detiber,

> Is this something that you believe needs to be solved by the initial implementation? Or is it something that we can likely add through iteration later?

This is a case where I think it's important to call out how this approach works with on-prem environments that don't have the vertical solution for compute instance/network management like AWS does for example. It's already an issue for worker nodes -- no good way to assign static IPs via MachineDeployments -- and this new model just extends the issue to the control plane.

Additionally, consider load balancers. CAPA has a cheat for LBs - it doesn't need to know the IPs of instances to connect an LB's target groups to specific instances, AWS ELBs can connect to instance refs before ever knowing the IP of an instance.

For CAPV we must know the IPs to connect LBs with the tech currently available. And this is likely true for other on-prem providers that exist today or might exist. When it comes to bringing your own LB ahead of time, this is possible because we can simply pre-define the LB's targets with static IPs and ensure the control plane machines are configured to use these static IPs. With this new control plane model, we'd be limited to DHCP addressing, making it nearly impossible to connect an existing load balancer with any type of acceptable UX.

dlipovetsky replied:
@akutz Thanks for bringing this up. I agree strongly that on-prem is an important use case, and that we should enable it as much as possible.

As for updating an external LB, I think there are a couple options that work with the current proposal . 
1. Have each Machine update the LB during bootstrapping; that's possible by adding pre/post commands to the KubeadmConfig.
2. Preallocate a pool of static IPs and configure the LB with them. The Infra Provider can follow label/annotation on a Machine to request an instance with an IP from the preallocated pool.

Sometimes an external LB is not available. Then a VIP might be used. That's a use case that should, I think, work with the default implementation, but maybe not in this first proposal.

randomvariable replied:
@thebsdbox any thoughts from your experimentations?

detiber replied:
I still want to be able to differentiate between nil and 0, since users might be confused if they explicitly set it to 0 and it magically becomes 1, vs nil becoming 1.

detiber replied:
Load balancer (or some other externally stable endpoint) has now been listed as a requirement.

ncdc commented:
Removing auto-approve so reviewers can manually indicate approval when it's ready.

/approve cancel

akutz commented:
Hi @detiber,

The conversation at https://github.com/kubernetes-sigs/cluster-api/pull/1613#discussion_r337582902 is no longer appearing because it's outdated, but my concern remains. I don't want to see this swept under a rug as an implementation detail without a discussion about how this *might* be handled, by CAPI or at the infrastructure level. 

I don't think a model should be accepted that assumes certain capabilities not shared by all interested and invested infrastructure providers.

akutz commented:
Hi @detiber,

The conversation at https://github.com/kubernetes-sigs/cluster-api/pull/1613#discussion_r337582902 is no longer appearing because it's outdated, but my concern remains. I don't want to see this swept under a rug as an implementation detail without a discussion about how this *might* be handled, by CAPI or at the infrastructure level. 

I don't think a model should be accepted that assumes certain capabilities not shared by all interested and invested infrastructure providers.

I'm going to add a hold until we can get a few people together to have a conversation about this topic. I'm not predicating the removal of the hold on an outcome that I find favorable, but just that we *have* the conversation at all. Thanks!

/hold

detiber commented:
@akutz I'm considering any non-resolved comments to be related to issues that are yet to be addressed (even if they appear as outdated).

akutz commented:
> @akutz I'm considering any non-resolved comments to be related to issues that are yet to be addressed (even if they appear as outdated).

Hi @detiber,

Thank you.

The link was as much for myself as anyone. I could literally not find the conversation when using Safari. I had to use Chrome and an e-mail link that took me directly to the comment. Safari wouldn't load it, nor did it show up.

rsmitty commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,564 @@
+)
+
+// ControlPlaneSpec defines the desired state of ControlPlane
+type ControlPlaneSpec struct {
```
We proposed this in slack, but may be worth doing a `s/ControlPlaneSpec/KubeadmControlPlaneSpec/` to make it clear that this is not a generic control plane resource, but will be kubeadm specific.

CecileRobertMichon replied:
+1

detiber replied:
Renaming has been done, thanks for the feedback!

CecileRobertMichon commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,564 @@
+
+// ControlPlaneSpec defines the desired state of ControlPlane
+type ControlPlaneSpec struct {
+    // Number of desired machines. Defaults to 1. When stacked etcd is used only 1, 3, 5, or 7 will be permitted, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members).
```
why 7 as the max?

dlipovetsky replied:
There is no hard maximum, but as the quorum increases, performance decreases, so most common deployments are 3 (to tolerate 1 failed member) and 5 (to tolerate 2 failed members).

I would be ok with allowing any odd number.

detiber replied:
Updated.

CecileRobertMichon commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,564 @@
+  sshKeyName: my-ssh-key
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+kind: AWSCluster
```
Would it make sense to make this more general? Maybe we can use a "fake" provider name like `SomeProviderCluster` since this is in the cluster-api repo we should avoid specific  infra-provider references in proposals IMO (unless we're giving examples for the other providers too). I know it's meant to be an example for illustration purposes but using a specific reference doesn't encourage the reader to have other infra providers in mind when reading the proposal.

dlipovetsky replied:
Good point. The concrete schema is helpful for discussion, but we could easily replace some terms (e.g. AWS, ami, region), but keep the schema.

detiber replied:
This has now been updated, thanks for the feedback.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+  - Pod based control planes; these can be managed via standard kubernetes objects.
+  - Non-node control planes (i.e. EKS, GKE, AKS); these can be managed via the respective APIs.
+- To manage CA certificates outside of what is provided by Kubeadm bootstrapping
+- To manage etcd clusters in any topology other than stacked etcd.
```
Correct, since we aren't involved with the lifecycle of the user-managed, external etcd cluster we do not care about how it is configured or managed :)

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+- To provide configuration of external cloud providers (i.e. the [cloud-controller-manager](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/)).This is deferred to kubeadm.
+- To provide CNI configuration. This is deferred to external, higher level tooling.
+- To provide the upgrade logic to handle changes to infrastructure (networks, firewalls etc…) that may need to be done to support a control plane on a newer version of Kubernetes (e.g. a cloud controller manager requires updated permissions against infrastructure APIs). We expect the work on add-on components to help to resolve some of these issues.
+- To provide horizontal or vertical auto scaling of control plane components, especially as etcd places hard performance limits beyond 3 nodes (due to latency).
```
The later, we are only addressing manual scaling (through declarative updates or the scale subresource).

detiber replied:
I tried rewording this, please let me know if it helps clarify.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+7. As a cluster operator, I want to be able to quickly respond to a Kubernetes CVE by upgrading my clusters in an automated fashion.
+8. As a cluster operator, I want to be able to quickly respond to a non-Kubernetes CVE that affects my base image or Kubernetes dependencies by upgrading my clusters in an automated fashion.
+9. As a cluster operator, I would like to upgrade to a new minor version of Kubernetes so that my cluster remains supported.
+10. As a cluster operator, I want to know why my cluster isn’t working properly after creation. I have ended up with an API server I can access, but CNI isn’t working, and new machines are not registering themselves with the control plane.
```
I think we can likely reword this, but the idea is to surface some type of a signal that there is an issue with the ControlPlane.

detiber replied:
I attempted to reword.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+#### etcd membership
+
+- If the leader is selected for deletion during a replacement for upgrade or scale down, the etcd cluster will be unavailable during that period as leader election takes place. Small time periods of unavailability should not significantly impact the running of the managed cluster’s API server.
+- Replication of the etcd log, if done for a sufficiently large datastore and saturates the network, machines may fail leader election, bringing down the cluster. To mitigate this, the control plane provider will only create machines serially, ensuring cluster health before moving onto operations for the next machine.
```
Correct.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+- To support auto remediation. We will treat control plane health as either informational or blocking for upgrades in the first instance based on real world feedback regarding a new control plane health indicator. Using such a mechanism to automatically replace machines may lead to unintended behaviours.
+- To provide configuration of external cloud providers (i.e. the [cloud-controller-manager](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/)).This is deferred to kubeadm.
+- To provide CNI configuration. This is deferred to external, higher level tooling.
+- To provide the upgrade logic to handle changes to infrastructure (networks, firewalls etc…) that may need to be done to support a control plane on a newer version of Kubernetes (e.g. a cloud controller manager requires updated permissions against infrastructure APIs). We expect the work on add-on components to help to resolve some of these issues.
```
done

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+### User Stories
+
+1. As a cluster operator, I want my Kubernetes clusters to have high availability control planes to meet my SLOs with application developers.
+2. As a developer, I want to be able to deploy as small as possible a cluster, e.g. to meet my organization’s cost requirements.
```
done.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+          cloud-provider: acme
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha2
+kind: AcmeMachineTemplate
```
Thanks for catching that! Fixed.

jaypipes commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+- To provide configuration of external cloud providers (i.e. the [cloud-controller-manager](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/)).This is deferred to kubeadm.
+- To provide CNI configuration. This is deferred to external, higher level tooling.
+- To provide the upgrade logic to handle changes to infrastructure (networks, firewalls etc…) that may need to be done to support a control plane on a newer version of Kubernetes (e.g. a cloud controller manager requires updated permissions against infrastructure APIs). We expect the work on [add-on components](https://github.com/kubernetes/community/tree/master/sig-cluster-lifecycle#cluster-addo ...
+- To provide automation around the horizontal or vertical scaling of control plane components, especially as etcd places hard performance limits beyond 3 nodes (due to latency).
```
@detiber yup, wording is much clearer now, ++

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,494 @@
+
+#### Identified features from user stories
+
+1. Based on the function of kubeadm, the control plane provider must be able to scale the number of replicas of a control plane from 1 to X, meeting user stories 1 through 4.
```
I'm rewording the first story to say "multiple control plane machines" and adding handling of failure domains as a non-goal (with a link to the issue tracking additional support for adding it).

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+    // ErrorMessage indicates that there is a problem reconciling the
+    // state, and will be set to a descriptive error message.
+    // +optional
+    ErrorMessage *string `json:"errorMessage,omitempty"`
```
We are deferring introduction of Conditions, since there isn't a clear consensus on how to implement them yet, we are tracking this in the following issue: https://github.com/kubernetes-sigs/cluster-api/issues/1658

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,566 @@
+    ErrorMessage *string `json:"errorMessage,omitempty"`
+
+    // Phase represents the current phase of control plane actuation.
+    // E.g. Pending, Initialized, Ready, ScalingUp, ScalingDown, Failed etc.
```
I've removed Phases, since we've already said that we want to move to more explictness around transitions of state and conditions here: https://github.com/kubernetes-sigs/cluster-api/issues/1658

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,569 @@
+  selector:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: my-cluster
+      cluster.x-k8s.io/controlplane: "true"
```
```suggestion
      cluster.x-k8s.io/control-plane: "true"
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,569 @@
+  - Configuration of external etcd is determined by introspecting the provided `KubeadmConfigSpec`.
+- `KubeadmControlPlane.Spec.Replicas` is >= 0 or is nil
+- `KubeadmControlPlane.Spec.Version != ""` (openapi)
+- `KubeadmControlPlane.Spec.KubeadmConfigSpec` must be treated as immutable (via webhook)
```
Would we want to allow modifications to `Pre/PostKubeadmCommands`, `Files`?

detiber replied:
Do you have specific use cases in mind? It might get a bit tricky to determine what are allowed vs not allowed modifications. We already capture the lack of mutability here in the Risks section, since once we adopt v1beta2 types for kubeadm there may need to be changes made to the kustomize templates provided.

ncdc replied:
Nothing specifc. Was trying to reconcile what it said below about triggering an update based on changes to this. I'm happy to start with it immutable, and loosen if we have feedback that it's needed.

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,569 @@
+
+##### Cluster upgrade (using create-swap-and-delete)
+
+- Triggered on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used.
```
It says above the kubeadm bits should be immutable? (although I just added a question there)

ncdc commented:
Added my GitHub approval

detiber commented:
/check-cla

vincepri commented:
/assign @timothysc 

for the last approval, before we can merge this proposal

akutz commented:
I thought reviewers were whomever provides feedback?

vincepri commented:
@akutz Yes, anyone can provide feedback or be added to the reviewer list. So far there is only 3 folks in the list.

https://github.com/kubernetes-sigs/cluster-api/pull/1613/files#diff-32997217ac55782fca9596683679c2bfR9-R11

Feel free to ask @detiber to be added, if you're interested in being added to the list!

akutz commented:
Hi @detiber,

Regarding https://github.com/kubernetes-sigs/cluster-api/pull/1613#discussion_r337582902, if a load balancer or some external endpoint is a requirement then doesn't that mean CAPI itself loses massive value for platforms without a built-in solution for an LB? @moshloop and I've been raising this issue in CAPI for a while, but it keeps getting punted as a "detail". However, the moment an external endpoint becomes a requirement for CAPI, it very possibly excludes existing infrastructure providers.

I cannot remove my hold on this CAEP until this is resolved.

akutz commented:
Hi @vincepri,

>  So far there is only 3 folks in the list.

I'm aware, which is why I asked the question. The one KEP I wrote, I took the task to keep the reviewers list updated to include people providing feedback. I thought that was the expectation. I was not aware I needed to ask. @detiber, may I pretty please be added to the list? :)

vincepri commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,563 @@
+  - "@ncdc"
+  - "@timothysc"
+  - "@vincepri"
+creation-date: 2019-09-18
```
Nit the creation date here doesn't reflect the file name

detiber commented:
@akutz my apologies for the oversight on the reviewer llist, I've now updated it.

alexbrand commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,568 @@
+- To manage etcd clusters in any topology other than stacked etcd.
+- To address disaster recovery constraints, e.g. restoring a control plane from 0 replicas using a filesystem or volume snapshot copy of data persisted in etcd.
+- To support rollbacks, as there is no datastore rollback guarantee for Kubernetes. Consumers should perform backups of the cluster prior to performing potentially destructive operations.
+- To mutate the configuration of live, running clusters (e.g. changing api-server flags), as this is the responsibility of the [component configuration working group](https://github.com/kubernetes/community/tree/master/wg-component-standard).
```
Will we have something in place to block this? We ran into a scenario where we needed to do this in v1alpha2. 

If we don't, it seems like it would be possible to change config flags by updating the `KubeadmConfigSpec` and (manually) rolling the control plane nodes one by one.

detiber replied:
@alexbrand currently we are adding a requirement that the `KubeadmConfigSpec` be treated immutably: https://github.com/kubernetes-sigs/cluster-api/pull/1613/files#diff-32997217ac55782fca9596683679c2bfR308

We've also called this out as part of the Risks: https://github.com/kubernetes-sigs/cluster-api/pull/1613/files#diff-32997217ac55782fca9596683679c2bfR542

alexbrand replied:
Gotcha! Thanks, @detiber. It seems like we'd need to create a new cluster if the KubeadmConfig required changes. I have to catch up on the ComponentConfig efforts as well.

detiber replied:
My main motivation here is to keep the initial implementation a bit simplified. We'll need to figure out how to detect "safe" vs "unsafe" changes in the future or a way to detect safe rollback for "configuration" or "base image" only changes where kubernetes version is not affected.

alexbrand replied:
Yes, absolutely. That makes sense to me 👍 .

alexbrand commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,568 @@
+- CA certificate secrets that were previously generated by the Kubeadm bootstrapper will now be generated by the KubeadmControlPlane Controller, maintaining backwards compatibility with the previous behavior if the KubeadmControlPlane is not used.
+- The kubeconfig secret that was previously created by the Machine Controller will now be generated by the KubeadmControlPlane Controller, maintaining backwards compatibility with the previous behavior if the KubeadmControlPlane is not used.
+
+#### Behaviors
```
Will Kubernetes Events be emitted to communicate progress to the user? Is it worth calling those out?

detiber replied:
I don't believe we've explicitly called out events in previous proposals and I was treating it as an implicit requirement, since we already emit events from the other controllers.

alexbrand replied:
Roger that 👍 
Thank you, @detiber.

ncdc commented:
Making a note here: need to add content described in https://github.com/kubernetes-sigs/cluster-api/issues/1687#issuecomment-548500168 to this proposal

sbueringer commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,576 @@
+- If an error occurs, `KubeadmControlPlane.Status.ErrorStatus` and `KubeadmControlPlane.Status.ErrorMessage` are populated.
+- `KubeadmControlPlane.Spec.Replicas` must be an odd number.
+- Can create an arbitrary number of control planes if etcd is external to the control plane, which will be determined by introspecting `KubeadmControlPlane.Spec.KubeadmConfigSpec`.
+- Creating a KubeadmControlPlane with > 1 replicas is equivalent to creating a KubeadmControlPlane with 1 replica followed by scaling the KubeadmControlPlane to 3 replicas
```
I assume you mean equivalent to create it with 1 replica and scaling it up to the specified count (not necessarily 3)

dlipovetsky replied:
Good catch, thanks!

detiber replied:
Fixed, thanks!

sbueringer commented on docs/proposals/images/controlplane/controlplane-init-7.plantuml:
```diff
@@ -0,0 +1,55 @@
+
+opt KubeadmControlPlane.Spec.Replicas >= 1, all existing Machines "Ready", num Machines > Replicas
+
+"KubeadmControlPlane Controller"->"API Server": Delete a Machine
```
Not sure which way is correct. In the text "An upgrade will look like this" has the following order:
* Remove the replica’s etcd member from the etcd cluster
* Update the kubeadm generated config map
* Delete the marked controlplane machine

But here it's the other way around. Do we first have to remove the etcd member or delete the machine?

I think first removing the etcd member is correct.

dlipovetsky replied:
> I think first removing the etcd member is correct.

I agree.

(When we delete a machine, the etcd process on that machine is stopped. We should remove the etcd member before stopping the etcd process; if we remove it afterward, there is a period of time when the etcd cluster has a failed member, as opposed to being a smaller, but healthy, cluster.)

randomvariable replied:
delete-then-replace was the original text, but was considered unsuitable for single-node control planes in this model, so was swapped. maybe we need both to address the trade-offs.

detiber replied:
There is another important drawback to the delete first scenario...  What happens to the API Servers that are configured for and attempting to use the etcd peer that has been removed from the etcd cluster? Will it properly fall back to the other configured etcd peers? Can we guarantee that behavior in the future?

I worry a bit that if we remove the member prior to deleting the instance, then we are opening ourselves up to undefined behavior for a period of time vs operating in a degraded state for short period of time until we remove the member.

That said, in either scenario, I don't think there is a significant difference in the risk related to the number of additional failures that could be tolerated by the etcd cluster for the time period we are discussing.

In an ideal scenario, we would be able to trigger a `kubeadm reset` on the control plane Machine prior to shutting down, which would be a better way to reduce the risk window for this action, but that would require quite a bit of care/architecture to ensure that we could implement it across the various known infrastruture provider implementations.

dlipovetsky replied:
> I worry a bit that if we remove the member prior to deleting the instance, then we are opening ourselves up to undefined behavior for a period of time vs operating in a degraded state for short period of time until we remove the member.

By "undefined behavior," do you mean we don't know what will happen if a control plane replica, otherwise running and healthy, has its etcd member removed unilaterally and externally via the ecd members API call (as opposed to running `kubeadm reset` on the replica)?

(From the docs and my own experiments, when an etcd member is removed, the corresponding etcd process terminates as soon as it knows about the change.)

> In an ideal scenario, we would be able to trigger a kubeadm reset on the control plane Machine prior to shutting down, which would be a better way to reduce the risk window for this action, but that would require quite a bit of care/architecture to ensure that we could implement it across the various known infrastruture provider implementations.

I agree.

detiber replied:
Fixed the ordering in the image, thanks!

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,576 @@
+  - Status.ControlPlaneReady is set based on the value of Status.Ready for the referenced resource, this field is intended to eventually replace Status.ControlPlaneInitialized as a field that will be kept up to date instead of set only once.
+- Current behavior will be preserved if `Cluster.Spec.ControlPlaneRef` is not set.
+- CA certificate secrets that were previously generated by the Kubeadm bootstrapper will now be generated by the KubeadmControlPlane Controller, maintaining backwards compatibility with the previous behavior if the KubeadmControlPlane is not used.
+- The kubeconfig secret that was previously created by the Machine Controller will now be generated by the KubeadmControlPlane Controller, maintaining backwards compatibility with the previous behavior if the KubeadmControlPlane is not used.
```
```suggestion
- The kubeconfig secret that was previously created by the Cluster Controller will now be generated by the KubeadmControlPlane Controller, maintaining backwards compatibility with the previous behavior if the KubeadmControlPlane is not used.
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,576 @@
+  selector:
+    matchLabels:
+      cluster.x-k8s.io/cluster-name: my-cluster
+      cluster.x-k8s.io/control-plane: "true"
```
```suggestion
      cluster.x-k8s.io/control-plane: ""
```

ncdc commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,576 @@
+  osImage:
+    id: objectstore-123456abcdef
+  instanceType: θ9.medium
+  iamInstanceProfile: "control-plane.cluster-api-provider-acme.sigs.k8s.io"
```
Do we want to use x-k8s.io here?

ncdc commented:
@hardikdr to date, we have not done a very good job of writing any behavioral contracts. I have opened #1727 to do this for control plane providers.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,576 @@
+During 2019 we saw control plane management implementations in each infrastructure provider. Much like
+bootstrapping was identified as being reimplemented in every infrastructure provider and then extracted into Cluster API
+Bootstrap Provider Kubeadm (CABPK), we believe we can reduce the redundancy of control plane management across providers
+and centralize the logic in Cluster API.
```
I added another sentence to support the enablement of alternative implementations.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,576 @@
+### Goals
+
+- To establish new resource types for control plane management
+- To support single node and multiple control plane instances, with the requirement that the infrastructure provider supports some type of a stable endpoint for the API Server (Load Balancer, VIP, etc).
```
Thanks, left off the hyphen, but made sure to use node consistently.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,576 @@
+Non-Goals listed in this document are intended to scope bound the current v1alpha3 implementation and are subject to change based on user feedback over time.
+
+- To manage non-machine based topologies, e.g.
+  - Pod based control planes; these can be managed via standard kubernetes objects.
```
Agreed, I had used previously suggested wording and have now updated it to remove the statements saying how such control planes should be implemented.

detiber commented on docs/proposals/20191017-control-plane.md:
```diff
@@ -0,0 +1,576 @@
+
+1. Based on the function of kubeadm, the control plane provider must be able to scale the number of replicas of a control plane from 1 to X, meeting user stories 1 through 4.
+2. To address user story 5, the control plane provider must provide validation of the number of replicas in a control plane. Where the stacked etcd topology is used (i.e., in the default implementation), the number of replicas must be an odd number, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members). When external etcd is used, any number is valid.
+3. In service of user story 5, the control plane provider must also manage etcd membership via kubeadm as part of scaling down (`kubeadm` takes care of adding the new etcd member when joining).
```
Updated to be explicit that this was talking about the kubeadm control plane implementation.

detiber commented:
@ncdc Added a current non-goal for the followup work related to stable api endpoints, we already have the current mechanism for using cluster.Status.APIEndpoints as part of the Constraints and Assumptions

detiber commented:
@hardikdr I believe I've addressed your inline comments.

I've also updated the non-goals to identify that documenting the exact interoperability mechanism for additional control plane providers will be handled as followup work and tracked by the issue that @ncdc linked.

detiber commented:
@chuckha @akutz @jaypipes @alexbrand @pablochacin @randomvariable @rsmitty @dlipovetsky @jzhoucliqr @sbueringer @CecileRobertMichon @hardikdr I believe all outstanding comments have been addressed. Please take a look, I'm hoping that we can go ahead and get this merged to unblock work to start on the initial implementation.

ncdc commented:
We would like to go with lazy consensus and plan to merge this on Wednesday as long as there are no blocking comments (will discuss at the community meeting)

moshloop commented:
> Uniquely mark the existing replica machine through setting an annotation and persisting the change (cluster.x-k8s.io/control-plane/selected-for-upgrade)

My only comment is to change this from an annotation to a label, so that it can be used in labelSelector

moshloop commented:
> Are you okay if we address this after merging as a followup issue?

Yes
/lgtm

detiber commented:
Created https://github.com/kubernetes-sigs/cluster-api/pull/1764 to address the feedback on the label vs annotation for upgrade selection.
